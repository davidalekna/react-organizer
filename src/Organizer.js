import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { days, months } from './utils';
import {
  format,
  getYear,
  getMonth,
  addMonths,
  subMonths,
  isSameMonth,
  isSameDay,
  eachDayOfInterval,
  addYears,
  subYears,
  isWeekend,
  setMonth,
  setYear,
} from 'date-fns';

const OrganizerContext = React.createContext({
  days: [],
  months: [],
  date: new Date(),
  selected: null,
  // functions
  getPrevMonthOffset: () => {},
  getNextMonthOffset: () => {},
  getCurrentMonth: () => {},
  getFullMonth: () => {},
  getFullYear: () => {},
  addCalendarMonth: () => {},
  subCalendarMonth: () => {},
  addCalendarYear: () => {},
  subCalendarYear: () => {},
  selectDate: () => {},
  reset: () => {},
  selectMonth: () => {},
  selectYear: () => {},
  changeLanguage: () => {},
});

export class Organizer extends React.Component {
  static propTypes = {
    children: PropTypes.func,
    initialDays: PropTypes.array,
    initialMonths: PropTypes.array,
    events: PropTypes.arrayOf(
      PropTypes.shape({
        starts: PropTypes.instanceOf(Date).isRequired,
      }),
    ),
    initialGridBlocks: PropTypes.number,
    initialDate: PropTypes.instanceOf(Date),
    initialSelected: PropTypes.instanceOf(Date),
  };
  static defaultProps = {
    stateReducer: (state, changes) => changes,
    onStateChange: () => {},
    onReset: () => {},
    onSelectDate: () => {},
    onAddCalendarYear: () => {},
    onSubCalendarYear: () => {},
    onSubCalendarMonth: () => {},
    onAddCalendarMonth: () => {},
    onSelectMonth: () => {},
    onSelectYear: () => {},
    onChangeLanguage: () => {},
    initialDays: days,
    initialMonths: months,
    events: [],
    initialGridBlocks: 41,
    initialDate: new Date(),
    initialSelected: null,
  };
  static stateChangeTypes = {
    reset: '__reset__',
    selectDate: '__select_date__',
    addCalendarMonth: '__add_calendar_month__',
    subCalendarMonth: '__subtract_calendar_month__',
    addCalendarYear: '__add_calendar_year__',
    subCalendarYear: '__sub_calendar_year__',
    selectMonth: '__select_month__',
    selectYear: '__select_year__',
    changeLanguage: '__change_language__',
  };
  static Consumer = OrganizerContext.Consumer;
  //
  _initializeEvents = forDate => {
    const getDate = d => format(d, 'dd/MM/yyyy');
    const currentDate = getDate(forDate);
    return this.props.events.filter(
      ({ starts }) => getDate(starts) === currentDate,
    );
  };
  _getWeeksInAMonth = (month, year) => {
    const weeks = [];
    const firstDate = new Date(year, month, 1);
    const lastDate = new Date(year, month + 1, 0);
    const numDays = lastDate.getDate();
    let start = 1;
    let end = 7 - firstDate.getDay();
    while (start <= numDays) {
      weeks.push({ start, end });
      start = end + 1;
      end = end + 7;
      if (end > numDays) end = numDays;
    }
    return weeks;
  };
  _getNumberOfDaysInAMonth(month, year) {
    return new Date(year, month + 1, 0).getDate();
  }
  changeLanguage = ({
    type = Organizer.stateChangeTypes.changeLanguage,
    days,
    months,
  }) => {
    if (days.length === 6 && months.length === 11) {
      this.internalSetState({ type, days, months }, () =>
        this.props.onChangeLanguage({
          days: this.getState().days,
          months: this.getState().months,
        }),
      );
    } else {
      throw new Error(
        `changeLanguage: Not enough days ${days.length} or months ${
          months.length
        }`,
      );
    }
  };
  getPrevMonthOffset = ({ month, year, events }) => {
    let prevMonthNumber = month - 2;
    let currentYear = year;
    if (prevMonthNumber < 0) {
      // check if previews year
      prevMonthNumber = 11;
      currentYear = currentYear - 1;
    }
    const { end, start } = this._getWeeksInAMonth(
      prevMonthNumber,
      currentYear,
    ).pop();
    let totalDays =
      this._getNumberOfDaysInAMonth(prevMonthNumber, currentYear) + 1;
    const assignDays = Array(end - start)
      .fill({})
      .map(() => {
        const currentDay = (totalDays -= 1);
        const date = new Date(currentYear, prevMonthNumber, currentDay);
        return {
          name: this.getState().days[date.getDay()],
          day: currentDay,
          date: date,
          offset: true,
          weekend: isWeekend(date),
          events: (events && this._initializeEvents(date)) || null,
        };
      })
      .reverse();
    return {
      name: this.getState().months[prevMonthNumber],
      month: prevMonthNumber + 1,
      year: currentYear,
      totalOffsetDays: assignDays.length,
      days: assignDays,
    };
  };
  getCurrentMonth = ({ month, year, events }) => {
    const currentMonth = month - 1;
    const totalDays = this._getNumberOfDaysInAMonth(currentMonth, year);
    const today = new Date().getDate() - 1;
    const selected = this.getState().date.getDate() - 1;
    const isToday =
      isSameMonth(new Date(year, currentMonth, today), new Date()) && today;
    const isSelected = calendarDay =>
      isSameDay(this.getState().selected, calendarDay);
    return {
      name: this.getState().months[currentMonth],
      month,
      year,
      totalDays,
      totalWeeks: this._getWeeksInAMonth(currentMonth, year).length,
      days: Array(totalDays)
        .fill({})
        .map((u, day) => {
          const currentDay = day + 1;
          const date = new Date(year, currentMonth, currentDay);
          return {
            name: this.getState().days[date.getDay()],
            day: currentDay,
            date: date,
            today: isToday === day,
            weekend: isWeekend(date),
            selected: isSelected(date),
            events: (events && this._initializeEvents(date)) || null,
          };
        }),
    };
  };
  getNextMonthOffset = ({
    month,
    year,
    totalOffsetDays,
    totalDays,
    events,
  }) => {
    let currentMonth = month;
    let currentYear = year;
    if (currentMonth > 11) {
      // check if next year
      currentMonth = 0;
      currentYear = currentYear + 1;
    }
    const nextMonthOffset =
      this.getState().gridBlocks - totalOffsetDays - totalDays;
    const assignDays = Array(nextMonthOffset)
      .fill({})
      .map((c, i) => {
        const currentDay = i + 1;
        const date = new Date(currentYear, currentMonth, currentDay);
        return {
          name: this.getState().days[date.getDay()],
          day: currentDay,
          date: date,
          offset: true,
          weekend: isWeekend(date),
          events: (events && this._initializeEvents(date)) || null,
        };
      });
    return {
      name: this.getState().months[currentMonth],
      month: currentMonth + 1,
      year: currentYear,
      totalOffsetDays: assignDays.length,
      days: assignDays,
    };
  };
  getFullMonth = (initMonth, events) => {
    const month = initMonth ? initMonth : getMonth(this.getState().date) + 1;
    const year = getYear(this.getState().date);
    const firstOffset = this.getPrevMonthOffset({ month, year, events });
    const current = this.getCurrentMonth({ month, year, events });
    const nextOffset = this.getNextMonthOffset({
      month,
      year,
      events,
      totalOffsetDays: firstOffset.totalOffsetDays,
      totalDays: current.totalDays,
    });
    return {
      ...current,
      days: [...firstOffset.days, ...current.days, ...nextOffset.days],
    };
  };
  getFullYear = events => {
    const months = Array(13)
      .fill({})
      .map((u, month) => this.getFullMonth(month, events));
    months.shift();
    return months;
  };
  addCalendarMonth = ({
    type = Organizer.stateChangeTypes.addCalendarMonth,
  }) => {
    this.internalSetState(
      state => ({
        type,
        date: addMonths(state.date, 1),
      }),
      () => this.props.onAddCalendarMonth(this.getState().date),
    );
  };
  selectMonth = ({ type = Organizer.stateChangeTypes.selectMonth, month }) => {
    this.internalSetState(
      state => ({
        type,
        date: setMonth(state.date, month),
      }),
      () => this.props.onSelectMonth(this.getState().date),
    );
  };
  subCalendarMonth = ({
    type = Organizer.stateChangeTypes.subCalendarMonth,
  }) => {
    this.internalSetState(
      state => ({
        type,
        date: subMonths(state.date, 1),
      }),
      () => this.props.onSubCalendarMonth(this.getState().date),
    );
  };
  addCalendarYear = ({ type = Organizer.stateChangeTypes.addCalendarYear }) => {
    this.internalSetState(
      state => ({
        type,
        date: addYears(state.date, 1),
      }),
      () => this.props.onAddCalendarYear(this.getState().date),
    );
  };
  selectYear = ({ type = Organizer.stateChangeTypes.selectYear, year }) => {
    this.internalSetState(
      state => ({
        type,
        date: setYear(state.date, year),
      }),
      () => this.props.onSelectYear(this.getState().date),
    );
  };
  subCalendarYear = ({ type = Organizer.stateChangeTypes.subCalendarYear }) => {
    this.internalSetState(
      state => ({
        type,
        date: subYears(state.date, 1),
      }),
      () => this.props.onSubCalendarYear(this.getState().date),
    );
  };
  selectDate = ({ type = Organizer.stateChangeTypes.selectDate, date }) => {
    this.internalSetState({ type, date, selected: date }, () => {
      return this.props.onSelectDate(this.getState().selected);
    });
  };
  reset = () => {
    this.internalSetState(
      { ...this.initialState, type: Organizer.stateChangeTypes.reset },
      () => this.props.onReset(this.getState().date),
    );
  };
  //
  initialState = {
    days: this.props.initialDays,
    months: this.props.initialMonths,
    gridBlocks: this.props.initialGridBlocks,
    date: this.props.initialDate,
    selected: this.props.initialSelected,
    // fns
    getPrevMonthOffset: this.getPrevMonthOffset,
    getNextMonthOffset: this.getNextMonthOffset,
    getCurrentMonth: this.getCurrentMonth,
    getFullMonth: this.getFullMonth,
    getFullYear: this.getFullYear,
    addCalendarMonth: this.addCalendarMonth,
    subCalendarMonth: this.subCalendarMonth,
    addCalendarYear: this.addCalendarYear,
    subCalendarYear: this.subCalendarYear,
    selectDate: this.selectDate,
    reset: this.reset,
    selectMonth: this.selectMonth,
    selectYear: this.selectYear,
    changeLanguage: this.changeLanguage,
  };
  state = this.initialState;
  isControlledProp(key) {
    return this.props[key] !== undefined;
  }
  getState(stateToMerge = this.state) {
    return Object.keys(stateToMerge).reduce((state, key) => {
      state[key] = this.isControlledProp(key)
        ? this.props[key]
        : stateToMerge[key];
      return state;
    }, {});
  }
  internalSetState = (changes, callback = () => {}) => {
    let allChanges;
    this.setState(
      currentState => {
        const combinedState = this.getState(currentState);
        return [changes]
          .map(c => (typeof c === 'function' ? c(currentState) : c))
          .map(c => {
            allChanges = this.props.stateReducer(combinedState, c) || {};
            return allChanges;
          })
          .map(({ type: ignoredType, ...onlyChanges }) => onlyChanges)
          .map(c => {
            return Object.keys(combinedState).reduce((newChanges, stateKey) => {
              if (!this.isControlledProp(stateKey)) {
                newChanges[stateKey] = c.hasOwnProperty(stateKey)
                  ? c[stateKey]
                  : combinedState[stateKey];
              }
              return newChanges;
            }, {});
          })
          .map(c => (Object.keys(c || {}).length ? c : null))[0];
      },
      () => {
        this.props.onStateChange(allChanges, this.state);
        callback();
      },
    );
  };
  render() {
    const { children } = this.props;
    const ui = typeof children === 'function' ? children(this.state) : children;
    return (
      <OrganizerContext.Provider value={this.state}>
        {ui}
      </OrganizerContext.Provider>
    );
  }
}

export function withOrganizer(Component) {
  const Wrapper = React.forwardRef((props, ref) => {
    return (
      <Organizer.Consumer>
        {browserUtils => (
          <Component {...props} organizer={browserUtils} ref={ref} />
        )}
      </Organizer.Consumer>
    );
  });
  Wrapper.displayName = `withOrganizer(${Component.displayName ||
    Component.name})`;
  hoistNonReactStatics(Wrapper, Component);
  return Wrapper;
}
