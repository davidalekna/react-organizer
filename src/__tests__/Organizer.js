import 'react-testing-library/cleanup-after-each';
import React from 'react';
import { render } from 'react-testing-library';
import Organizer from '../';

const initialDate = { month: 9, year: 2018 };

test('should get previews month last weeks offset days', () => {
  const { getPrevMonthOffset } = setup();
  const result = getPrevMonthOffset(initialDate);
  const offsetDay = result.days[0];

  expect(result.name).toEqual('August');
  expect(result.month).toEqual(8);
  expect(result.year).toEqual(2018);
  expect(result.totalOffsetDays).toEqual(5);
  expect(result.days.length).toEqual(5);
});

test('should get current month days and events', () => {
  const { getCurrentMonth } = setup();
  const result = getCurrentMonth(initialDate);
  const offsetDay = result.days[0];

  expect(result.name).toEqual('September');
  expect(result.month).toEqual(9);
  expect(result.year).toEqual(2018);
  expect(result.totalDays).toEqual(30);
  expect(result.totalWeeks).toEqual(6);
  expect(result.days.length).toEqual(30);
});

test('should get next month first weeks offset days', () => {
  const { getNextMonthOffset } = setup();
  const result = getNextMonthOffset({
    ...initialDate,
    totalOffsetDays: 5,
    totalDays: 30,
  });
  const offsetDay = result.days[0];
  expect(result.name).toEqual('October');
  expect(result.month).toEqual(10);
  expect(result.year).toEqual(2018);
  expect(result.totalOffsetDays).toEqual(7);
  expect(result.days.length).toEqual(7);
});

test('should get full month with prev offset days and next offset days', () => {
  const { getFullMonth } = setup();
  const result = getFullMonth(9, true);
  const offsetDays = result.days.filter(day => day.offset);
  expect(result.name).toEqual('September');
  expect(result.month).toEqual(9);
  expect(result.year).toEqual(2018);
  expect(result.totalDays).toEqual(30);
  expect(result.totalWeeks).toEqual(6);
  expect(result.days.length).toEqual(6 + 30 + 6);
  expect(offsetDays.length).toEqual(6 + 6);
});

test('should get full calendar year', () => {
  const { getFullYear } = setup();
  const result = getFullYear();
  expect(result.length).toEqual(12);
});

test('should add calendar month', () => {
  const handleStateChange = jest.fn();
  const { addCalendarMonth } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  addCalendarMonth();
  const changes = {
    type: Organizer.stateChangeTypes.addCalendarMonth,
    date: new Date('2018', '10', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('2018', '10', '17') }),
  );
});

test('should subtract calendar month', () => {
  const handleStateChange = jest.fn();
  const { subCalendarMonth } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  subCalendarMonth();
  const changes = {
    type: Organizer.stateChangeTypes.subCalendarMonth,
    date: new Date('2018', '08', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('2018', '08', '17') }),
  );
});

test('should add calendar year', () => {
  const handleStateChange = jest.fn();
  const { addCalendarYear } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  addCalendarYear();
  const changes = {
    type: Organizer.stateChangeTypes.addCalendarYear,
    date: new Date('2019', '09', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('2019', '09', '17') }),
  );
});

test('should subtract calendar year', () => {
  const handleStateChange = jest.fn();
  const { subCalendarYear } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  subCalendarYear();
  const changes = {
    type: Organizer.stateChangeTypes.subCalendarYear,
    date: new Date('2017', '09', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('2017', '09', '17') }),
  );
});

test('should select requested date', () => {
  const handleStateChange = jest.fn();
  const { selectDate } = setup({
    onStateChange: handleStateChange,
  });
  selectDate({ date: new Date('2022', '09', '17') });
  const changes = {
    type: Organizer.stateChangeTypes.selectDate,
    date: new Date('2022', '09', '17'),
    selected: new Date('2022', '09', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({
      date: new Date('2022', '09', '17'),
      selected: new Date('2022', '09', '17'),
    }),
  );
});

test('selectRange should select range', () => {
  const handleStateChange = jest.fn();
  const { selectRange } = setup({
    onStateChange: handleStateChange,
  });
  selectRange({ date: new Date('2022', '09', '17') });
  selectRange({ date: new Date('2022', '09', '22') });
  const changes = {
    type: Organizer.stateChangeTypes.selectRange,
    date: new Date('2022', '09', '22'),
    selected: [new Date('2022', '09', '17'), new Date('2022', '09', '22')],
  };

  expect(handleStateChange).toHaveBeenCalledTimes(2);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({
      date: new Date('2022', '09', '22'),
      selected: [new Date('2022', '09', '17'), new Date('2022', '09', '22')],
    }),
  );
});

test('should select requested month', () => {
  const handleStateChange = jest.fn();
  const { selectMonth } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  selectMonth({ month: 11 });
  const changes = {
    type: Organizer.stateChangeTypes.selectMonth,
    date: new Date('2018', '11', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('2018', '11', '17') }),
  );
});

test('should select requested year', () => {
  const handleStateChange = jest.fn();
  const { selectYear } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  selectYear({ year: 1989 });
  const changes = {
    type: Organizer.stateChangeTypes.selectYear,
    date: new Date('1989', '09', '17'),
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ date: new Date('1989', '09', '17') }),
  );
});

test('should change language', () => {
  const handleStateChange = jest.fn();
  const { changeLanguage } = setup({
    initialDate: new Date('2018', '09', '17'),
    onStateChange: handleStateChange,
  });
  const days = [
    'понедельник',
    'вторник',
    'среда',
    'четверг',
    'пятница',
    'суббота',
    'воскресенье',
  ];
  const months = [
    'январь',
    'февраль',
    'март',
    'апрель',
    'май',
    'июнь',
    'июль',
    'август',
    'сентябрь',
    'октябрь',
    'ноябрь',
    'декабрь',
  ];
  changeLanguage({ days, months });
  const changes = {
    type: Organizer.stateChangeTypes.changeLanguage,
    days,
    months,
  };

  expect(handleStateChange).toHaveBeenCalledTimes(1);
  expect(handleStateChange).toHaveBeenLastCalledWith(
    changes,
    expect.objectContaining({ days, months }),
  );
});

// test('should reset the state', () => {
//   const handleStateChange = jest.fn();
//   const { reset } = setup({
//     onStateChange: handleStateChange,
//   });
//   reset();
//   const changes = Organizer.initialState;

//   expect(handleStateChange).toHaveBeenCalledTimes(1);
//   expect(handleStateChange).toHaveBeenLastCalledWith(
//     changes,
//     expect.objectContaining(Organizer.initialState),
//   );
// });

function setup({ render: renderFn = () => <div />, ...props } = {}) {
  let renderArg;
  const childrenSpy = jest.fn(controllerArg => {
    renderArg = controllerArg;
    return renderFn(controllerArg);
  });
  const utils = render(<Organizer {...props}>{childrenSpy}</Organizer>);
  return { childrenSpy, ...utils, ...renderArg };
}
