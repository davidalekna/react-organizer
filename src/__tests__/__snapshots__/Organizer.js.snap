// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get current month days and events 1`] = `
Object {
  "date": 2018-08-31T23:00:00.000Z,
  "day": 1,
  "events": null,
  "name": "Sunday",
  "selected": false,
  "today": false,
  "weekend": true,
}
`;

exports[`should get full month with prev offset days and next offset days 1`] = `
Array [
  Object {
    "date": 2018-08-26T23:00:00.000Z,
    "day": 27,
    "events": Array [],
    "name": "Tuesday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-08-27T23:00:00.000Z,
    "day": 28,
    "events": Array [],
    "name": "Wednesday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-08-28T23:00:00.000Z,
    "day": 29,
    "events": Array [],
    "name": "Thursday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-08-29T23:00:00.000Z,
    "day": 30,
    "events": Array [],
    "name": "Friday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-08-30T23:00:00.000Z,
    "day": 31,
    "events": Array [],
    "name": "Saturday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-09-30T23:00:00.000Z,
    "day": 1,
    "events": Array [],
    "name": "Tuesday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-10-01T23:00:00.000Z,
    "day": 2,
    "events": Array [],
    "name": "Wednesday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-10-02T23:00:00.000Z,
    "day": 3,
    "events": Array [],
    "name": "Thursday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-10-03T23:00:00.000Z,
    "day": 4,
    "events": Array [],
    "name": "Friday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-10-04T23:00:00.000Z,
    "day": 5,
    "events": Array [],
    "name": "Saturday",
    "offset": true,
    "weekend": false,
  },
  Object {
    "date": 2018-10-05T23:00:00.000Z,
    "day": 6,
    "events": Array [],
    "name": "Sunday",
    "offset": true,
    "weekend": true,
  },
]
`;

exports[`should get next month first weeks offset days 1`] = `
Object {
  "date": 2018-09-30T23:00:00.000Z,
  "day": 1,
  "events": null,
  "name": "Tuesday",
  "offset": true,
  "weekend": false,
}
`;

exports[`should get previews month last weeks offset days 1`] = `
Object {
  "date": 2018-08-26T23:00:00.000Z,
  "day": 27,
  "events": null,
  "name": "Tuesday",
  "offset": true,
  "weekend": false,
}
`;
